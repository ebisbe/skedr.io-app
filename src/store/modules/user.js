import Auth from '@aws-amplify/auth'
import LogRocket from 'logrocket'

export const state = {
  user: null,
  username: null,
  userId: null,
  userVerification: null
}

export const getters = {
  userId: state => state.username
}

export const mutations = {
  setUser(state, user) {
    state.user = user
    if (user !== null) {
      state.username = decodeURIComponent(user.username)
    } else {
      state.username = null
    }
  },
  setUserAttributes(state, attributes) {
    state.user.attributes = Object.assign({}, state.user.attributes, attributes)
  },
  setUserId: (state, userId) => (state.userId = userId),
  setUserVerification: (state, data) => {
    state.userVerification = data
  }
}

export const actions = {
  loginUser: async function({ state, commit }, { username, password }) {
    // the value of `username` is really an `email`
    const user = await Auth.signIn(username.toLowerCase(), password)
    commit('setUser', user)

    const data = await Auth.verifiedContact(user)
    commit('setUserVerification', data)

    const payload = {
      body: { userId: state.username, email: username }
    }
    return Object.keys(data.verified).length !== 0
  },
  currentAuthenticatedUser: async ({ commit, state }) => {
    try {
      const user = await Auth.currentAuthenticatedUser({ bypassCache: true })
      const credentials = await Auth.currentCredentials()
      commit('setUser', user)
      commit('setUserId', credentials.identityId)
      LogRocket.identify(state.username, {
        name: `${user.attributes.name} ${user.attributes.family_name}`,
        email: user.attributes.email
      })
      LogRocket.getSessionURL(sessionURL => {
        const indentifyUser = {
          name: `${user.attributes.name} ${user.attributes.family_name}`,
          email: user.attributes.email,
          sessionURL: sessionURL
        }
        Beacon('identify', indentifyUser)
      })
    } catch (err) {
      commit('setUser', null)
      return false
    }
    return true
  },
  updateUserAttributes: async ({ commit, state }) => {
    await Auth.updateUserAttributes(state.user, state.user.attributes)
    const user = await Auth.currentAuthenticatedUser({ bypassCache: true })
    commit('setUser', user)
  },
  confirmEmailCode: async ({ commit }, code) => {
    await Auth.verifyCurrentUserAttributeSubmit('email', code)
    const user = await Auth.currentAuthenticatedUser({ bypassCache: true })
    commit('setUser', user)
  },
  signOut: async ({ state }) => {
    if (state.user.attributes.email_verified === false) {
      throw {
        name: 'VerifyEmailError',
        message: 'To sign out first verify your email'
      }
    }
    await Auth.signOut()
    return true
  }
}
