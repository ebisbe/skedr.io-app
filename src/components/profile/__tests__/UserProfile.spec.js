import { shallowMount } from '@vue/test-utils'
import Vuetify from 'vuetify'
import Vue from 'vue'
import Vuex from 'vuex'
import comp from '../UserProfile'
import { mutations } from '@/store/modules/user'

Vue.use(Vuetify)
Vue.use(Vuex)

const createStore = ({ email_verified = true, actions = {} } = {}) =>
  new Vuex.Store({
    modules: {
      user: {
        namespaced: true,
        state: {
          user: {
            attributes: {
              name: 'Enric',
              family_name: 'Bisbe Gil',
              email: 'email@domain.com',
              email_verified
            }
          }
        },
        mutations,
        actions
      }
    }
  })

const createComp = dataStore =>
  shallowMount(comp, {
    store: createStore(dataStore)
  })

describe('UserProfile', () => {
  it('tests the snapshot with email verified', () => {
    const wrapper = createComp()
    expect(wrapper).toMatchSnapshot()
  })
  it('tests the snapshot without the email verified', () => {
    const wrapper = createComp({ email_verified: false })
    expect(wrapper).toMatchSnapshot()
  })

  describe('Computed properties', () => {
    it('sets name', () => {
      const wrapper = createComp()
      expect(wrapper.vm.disableUploadBtn).toBe(true)
      wrapper.vm.name = 'Name'
      expect(wrapper.vm.name).toBe('Name')
      expect(wrapper.vm.disableUploadBtn).toBe(false)
    })
    it('sets family name', () => {
      const wrapper = createComp()
      expect(wrapper.vm.disableUploadBtn).toBe(true)
      wrapper.vm.family_name = 'Family Name'
      expect(wrapper.vm.family_name).toBe('Family Name')
      expect(wrapper.vm.disableUploadBtn).toBe(false)
    })
    it('sets email', () => {
      const wrapper = createComp()
      expect(wrapper.vm.disableUploadBtn).toBe(true)
      wrapper.vm.email = 'email'
      expect(wrapper.vm.email).toBe('email')
      expect(wrapper.vm.disableUploadBtn).toBe(false)
    })
    it('opens the dialog', () => {
      const wrapper = createComp({ email_verified: false })
      expect(wrapper.vm.openDialog).toBe(true)
    })
    it('validates code is a six value number', () => {
      const wrapper = createComp()
      wrapper.vm.code = '123456'
      expect(wrapper.vm.isCodeValid).toBe(true)
    })
  })
  describe('Methods', () => {
    it('updates a user attributes', async () => {
      const updateUserAttributes = jest.fn()
      const wrapper = createComp({ actions: { updateUserAttributes } })
      wrapper.vm.code = '123456'
      wrapper.vm.disableUploadBtn = false

      await wrapper.vm.updateUser()

      expect(updateUserAttributes).toHaveBeenCalledTimes(1)
      expect(wrapper.vm.code).toBe('')
      expect(wrapper.vm.disableUploadBtn).toBe(true)
      expect(wrapper.vm.loadingBtn).toBe(false)
      expect(wrapper.vm.closeDialog).toBe(false)
    })
    it('confirms email', async () => {
      const confirmEmailCode = jest.fn()
      const wrapper = createComp({ actions: { confirmEmailCode } })

      await wrapper.vm.confirmEmail()

      expect(confirmEmailCode).toHaveBeenCalledTimes(1)
      expect(wrapper.vm.loadingBtn).toBe(false)
      expect(wrapper.vm.closeDialog).toBe(false)
    })
  })
})
