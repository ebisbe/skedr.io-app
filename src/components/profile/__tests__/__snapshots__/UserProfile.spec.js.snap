// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfile tests the snapshot with email verified 1`] = `
<v-card-stub loaderheight="4" tag="div" class="mb-6">
  <v-form-stub>
    <v-container-stub tag="div">
      <h1></h1>
      <v-text-field-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" value="Enric" backgroundcolor="" label="label.first_name" loaderheight="2" clearicon="$clear" type="text"></v-text-field-stub>
      <v-text-field-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" value="Bisbe Gil" backgroundcolor="" label="label.family_name" loaderheight="2" clearicon="$clear" type="text"></v-text-field-stub>
      <v-text-field-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" value="email@domain.com" backgroundcolor="" label="label.email" loaderheight="2" clearicon="$clear" type="text"></v-text-field-stub>
      <v-card-actions-stub class="px-6">
        <v-spacer-stub></v-spacer-stub>
        <!---->
        <v-btn-stub color="primary" tag="button" activeclass="" disabled="true" type="button"></v-btn-stub>
      </v-card-actions-stub>
    </v-container-stub>
    <v-dialog-stub opendelay="0" closedelay="0" contentclass="" maxwidth="500" origin="center center" persistent="true" retainfocus="true" transition="dialog-transition" width="auto">
      <v-card-stub loaderheight="4" tag="div" class="elevation-12">
        <v-toolbar-stub color="primary" tag="header" extensionheight="48" src="">
          <v-toolbar-title-stub></v-toolbar-title-stub>
        </v-toolbar-stub>
        <v-card-text-stub class="px-6 pb-6 pt-4">
          UserProfile.code_sent
          <v-text-field-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" value="" backgroundcolor="" label="UserProfile.confirmation_code" loaderheight="2" clearicon="$clear" counter="6" type="text"></v-text-field-stub>
          <v-card-actions-stub class="px-6">
            <v-spacer-stub></v-spacer-stub>
            <v-btn-stub tag="button" activeclass="" text="true" type="button"></v-btn-stub>
            <v-btn-stub color="primary" tag="button" activeclass="" disabled="true" type="button"></v-btn-stub>
          </v-card-actions-stub>
        </v-card-text-stub>
      </v-card-stub>
    </v-dialog-stub>
  </v-form-stub>
</v-card-stub>
`;

exports[`UserProfile tests the snapshot without the email verified 1`] = `
<v-card-stub loaderheight="4" tag="div" class="mb-6">
  <v-form-stub>
    <v-container-stub tag="div">
      <h1></h1>
      <v-text-field-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" value="Enric" backgroundcolor="" label="label.first_name" loaderheight="2" clearicon="$clear" type="text"></v-text-field-stub>
      <v-text-field-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" value="Bisbe Gil" backgroundcolor="" label="label.family_name" loaderheight="2" clearicon="$clear" type="text"></v-text-field-stub>
      <v-text-field-stub errorcount="1" errormessages="UserProfile.error_verify_email" messages="" rules="" successmessages="" value="email@domain.com" backgroundcolor="" label="label.email" loaderheight="2" clearicon="$clear" type="text"></v-text-field-stub>
      <v-card-actions-stub class="px-6">
        <v-spacer-stub></v-spacer-stub>
        <v-btn-stub color="primary" tag="button" activeclass="" type="button"></v-btn-stub>
        <v-btn-stub color="primary" tag="button" activeclass="" disabled="true" type="button"></v-btn-stub>
      </v-card-actions-stub>
    </v-container-stub>
    <v-dialog-stub opendelay="0" closedelay="0" value="true" contentclass="" maxwidth="500" origin="center center" persistent="true" retainfocus="true" transition="dialog-transition" width="auto">
      <v-card-stub loaderheight="4" tag="div" class="elevation-12">
        <v-toolbar-stub color="primary" tag="header" extensionheight="48" src="">
          <v-toolbar-title-stub></v-toolbar-title-stub>
        </v-toolbar-stub>
        <v-card-text-stub class="px-6 pb-6 pt-4">
          UserProfile.code_sent
          <v-text-field-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" value="" backgroundcolor="" label="UserProfile.confirmation_code" loaderheight="2" clearicon="$clear" counter="6" type="text"></v-text-field-stub>
          <v-card-actions-stub class="px-6">
            <v-spacer-stub></v-spacer-stub>
            <v-btn-stub tag="button" activeclass="" text="true" type="button"></v-btn-stub>
            <v-btn-stub color="primary" tag="button" activeclass="" disabled="true" type="button"></v-btn-stub>
          </v-card-actions-stub>
        </v-card-text-stub>
      </v-card-stub>
    </v-dialog-stub>
  </v-form-stub>
</v-card-stub>
`;
